File: jamster.engine/Events/Penalties.cs
using jamster.engine.Domain;
using jamster.engine.Services;

namespace jamster.engine.Events;

public sealed class SkaterSatInBox(Guid7 id, SkaterSatInBoxBody body) : Event<SkaterSatInBoxBody>(id, body);
public sealed record SkaterSatInBoxBody(TeamSide TeamSide, Guid SkaterId) : TeamEventBody(TeamSide);

public sealed class SkaterReleasedFromBox(Guid7 id, SkaterReleasedFromBoxBody body) : Event<SkaterReleasedFromBoxBody>(id, body);
public sealed record SkaterReleasedFromBoxBody(TeamSide TeamSide, Guid SkaterId) : TeamEventBody(TeamSide);

public sealed class SkaterSubstitutedInBox(Guid7 id, SkaterSubstitutedInBoxBody body) : Event<SkaterSubstitutedInBoxBody>(id, body);
public sealed record SkaterSubstitutedInBoxBody(TeamSide TeamSide, Guid OriginalSkaterId, Guid NewSkaterId) : TeamEventBody(TeamSide);

public sealed class PenaltyAssessed(Guid7 id, PenaltyAssessedBody body) : Event<PenaltyAssessedBody>(id, body);
public sealed record PenaltyAssessedBody(TeamSide TeamSide, Guid SkaterId, string PenaltyCode) : TeamEventBody(TeamSide);

public sealed class PenaltyRescinded(Guid7 id, PenaltyRescindedBody body) : Event<PenaltyRescindedBody>(id, body);
public sealed record PenaltyRescindedBody(TeamSide TeamSide, Guid SkaterId, string PenaltyCode, int Period, int Jam) : TeamEventBody(TeamSide);

public sealed class PenaltyUpdated(Guid7 id, PenaltyUpdatedBody body) : Event<PenaltyUpdatedBody>(id, body);
public sealed record PenaltyUpdatedBody(TeamSide TeamSide, Guid SkaterId, string OriginalPenaltyCode, int OriginalPeriod, int OriginalJam, string NewPenaltyCode, int NewPeriod, int NewJam) : TeamEventBody(TeamSide);

public sealed class SkaterExpelled(Guid7 id, SkaterExpelledBody body) : Event<SkaterExpelledBody>(id, body);
public sealed record SkaterExpelledBody(TeamSide TeamSide, Guid SkaterId, string PenaltyCode, int Period, int Jam) : TeamEventBody(TeamSide);

public sealed class ExpulsionCleared(Guid7 id, ExpulsionClearedBody body) : Event<ExpulsionClearedBody>(id, body);
public sealed record ExpulsionClearedBody(TeamSide TeamSide, Guid SkaterId) : TeamEventBody(TeamSide);

public sealed class PenaltyServedSet(Guid7 id, PenaltyServedSetBody body) : Event<PenaltyServedSetBody>(id, body);
public sealed record PenaltyServedSetBody(TeamSide TeamSide, Guid SkaterId, string PenaltyCode, int Period, int Jam, bool Served) : TeamEventBody(TeamSide);

File: jamster.ui/src/types/events/Penalties.ts
import { EventWithBody } from "@/hooks";
import { TeamSide } from "@/types";

export class SkaterSatInBox extends EventWithBody {
    constructor(teamSide: TeamSide, skaterId: string) {
        super("SkaterSatInBox", { teamSide, skaterId });
    }
}

export class SkaterReleasedFromBox extends EventWithBody {
    constructor(teamSide: TeamSide, skaterId: string) {
        super("SkaterReleasedFromBox", { teamSide, skaterId });
    }
}

export class SkaterSubstitutedInBox extends EventWithBody {
    constructor(teamSide: TeamSide, originalSkaterId: string, newSkaterId: string) {
        super("SkaterSubstitutedInBox", { teamSide, originalSkaterId, newSkaterId });
    }
}

export class PenaltyAssessed extends EventWithBody {
    constructor(teamSide: TeamSide, skaterId: string, penaltyCode: string) {
        super("PenaltyAssessed", { teamSide, skaterId, penaltyCode });
    }
}

export class PenaltyRescinded extends EventWithBody {
    constructor(teamSide: TeamSide, skaterId: string, penaltyCode: string, period: number, jam: number) {
        super("PenaltyRescinded", { teamSide, skaterId, penaltyCode, period, jam });
    }
}

export class PenaltyUpdated extends EventWithBody {
    constructor(teamSide: TeamSide, skaterId: string, originalPenaltyCode: string, originalPeriod: number, originalJam: number, newPenaltyCode: string, newPeriod: number, newJam: number) {
        super("PenaltyUpdated", {
            teamSide,
            skaterId,
            originalPenaltyCode,
            originalPeriod,
            originalJam,
            newPenaltyCode,
            newPeriod,
            newJam
        });
    }
}

export class SkaterExpelled extends EventWithBody {
    constructor(teamSide: TeamSide, skaterId: string, penaltyCode: string, period: number, jam: number) {
        super("SkaterExpelled", { teamSide, skaterId, penaltyCode, period, jam });
    }
}

export class ExpulsionCleared extends EventWithBody {
    constructor(teamSide: TeamSide, skaterId: string) {
        super("ExpulsionCleared", { teamSide, skaterId });
    }
}

export class PenaltyServedSet extends EventWithBody {
    constructor(teamSide: TeamSide, skaterId: string, penaltyCode: string, period: number, jam: number, served: boolean) {
        super("PenaltyServedSet", { teamSide, skaterId, penaltyCode, period, jam, served });
    }
}